e2e: bring back total test spec for Ginkgo v2

chendave opened the pull request:
/kind bug



<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/114313

#### Special notes for your reviewer:
this was tested locally,

```bash
dlv exec ./e2e.test -- -ginkgo.focus="NoExecuteTaintManager Single Pod \[Serial\]" -report-dir=/var/log/kubernetes/e2e -spec-dump=/var/log/kubernetes/dump

p reporter
*k8s.io/kubernetes/test/e2e/reporters.ProgressReporter {
        LastMsg: "PASSED [sig-node] NoExecuteTaintManager Single Pod [Serial] even...+59 more",
        TestsTotal: 5,
        TestsCompleted: 5,
        TestsSkipped: 7067,
        TestsFailed: 0,
        Failures: []string len: 0, cap: 0, [],
        progressURL: "",
        client: *net/http.Client nil,}
(dlv) p reporter.TestsTotal
5

```


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
total test spec is now available by `ProgressReporter`, it will be reported before test suite got executed.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```


/sig testing

chendave commented:
/cc @pohly @aojea @onsi  @dims @raosaurabh

pohly commented:
Don't you have to use `ReportBeforeSuite` somewhere?

chendave commented:
> Don't you have to use ReportBeforeSuite somewhere?

It's here:
https://github.com/kubernetes/kubernetes/pull/114417/commits/ed062c13529fa547e86acf6a5b69a5c340400b94#diff-d5d12b6ea07b0a78c9ad1987e1486c565ecf16839f77a93e91b09475cf6c0e99R153-R156

aojea commented:
> > Don't you have to use ReportBeforeSuite somewhere?
> 
> It's here: [ed062c1#diff-d5d12b6ea07b0a78c9ad1987e1486c565ecf16839f77a93e91b09475cf6c0e99R153-R156](https://github.com/kubernetes/kubernetes/commit/ed062c13529fa547e86acf6a5b69a5c340400b94#diff-d5d12b6ea07b0a78c9ad1987e1486c565ecf16839f77a93e91b09475cf6c0e99R153-R156)

can you do the ginkgo bump and that reporters change in separate commits?

chendave commented:
> can you do the ginkgo bump and that reporters change in separate commits?

unfortunately not, `EmitSpecProgress` has been removed from `suiteConfig` since 2.5.0, unittest won't pass with that change.

chendave commented:
/hold
still need to check the new behavior of `show-node-events` option

pohly commented:
> > Don't you have to use ReportBeforeSuite somewhere?
> 
> It's here: [ed062c1#diff-d5d12b6ea07b0a78c9ad1987e1486c565ecf16839f77a93e91b09475cf6c0e99R153-R156](https://github.com/kubernetes/kubernetes/commit/ed062c13529fa547e86acf6a5b69a5c340400b94#diff-d5d12b6ea07b0a78c9ad1987e1486c565ecf16839f77a93e91b09475cf6c0e99R153-R156)

I would prefer to have a separate commit with that change.

chendave commented:
the failure of `TestCleanup` is valid, with the removal of `EmitSpecProgress`, we must update the string of `ginkgoOutput` to adapt with the new latest change.

onsi commented:
Sorry y'all that's my bad on removing the config field vs deprecating it.  I forgot that some folks are accessing the config object directly now.  I can cut a 2.6.1 that reintroduces it as a no-op if that would help here.

pohly commented:
We can adapt to such API changes, that's not a problem. The question is just how to organize the commits in the PR.

"Update ginkgo" in one commit followed by a "adapt to API changes" commit and finally a "use ReportBeforeSuite" commit is easiest to review. The downside is that bisecting breaks because the first commit by itself doesn't compile, but the same is true also for other PRs.

aojea commented:
> The downside is that bisecting breaks because the first commit by itself doesn't compile, but the same is true also for other PRs.

yeah, this is not new, is not the most elegant thing but it has worked fine for us until now ... otherwise finding the change on the commit with the revendoring is super complicated

pohly commented:
> the failure of TestCleanup is valid, with the removal of EmitSpecProgress, we must update the string of ginkgoOutput to adapt with the latest change.

I think I already did that in https://github.com/kubernetes/kubernetes/pull/113538. I'll update that PR to include the newer Ginkgo.

alexzielenski commented:
/remove-sig api-machinery

chendave commented:
I will rebase once https://github.com/kubernetes/kubernetes/pull/113538 got merged.

bart0sh commented:
/priority important-soon
/triage accepted

chendave committed:
e2e: bring back total test spec for `Ginkgo` v2

Bring back the number of test spec which was dropped earlier.
It's now available in the reporting node of `ReportBeforeSuite` by extracting
the number from report.PreRunStats.SpecsThatWillBeRun.

Signed-off-by: Dave Chen <dave.chen@arm.com>

chendave commented:
/retitle e2e: bring back total test spec for Ginkgo v2

chendave commented:
/hold cancel

rebase done!
@pohly @aojea @dims  this is ready for review now.

aojea commented on test/e2e/e2e_test.go:
```diff
@@ -151,6 +151,10 @@ var _ = ginkgo.ReportAfterEach(func(report ginkgo.SpecReport) {
 	progressReporter.ProcessSpecReport(report)
 })
 
+var _ = ginkgo.ReportBeforeSuite(func(report ginkgo.Report) {
```
I honestly can't remember the details, and don't have much time to dig into the code, so sorry for the questions:

does this means this reporter always run? or is still optional?
is it enabled or disabled by default in the CI?

chendave replied:
@aojea the report is only enabled for e2e test binary: https://github.com/kubernetes/kubernetes/blob/2bb77a13b1a709e0208c6ba9ba5323a5e54f79d6/test/e2e/e2e.go#L112

and it's not used anywhere else as far as I know.

the progress reporter can be treated as *optional* if end user doesn't set the `progress-report-url`, pls see:
https://github.com/kubernetes/kubernetes/blob/2bb77a13b1a709e0208c6ba9ba5323a5e54f79d6/test/e2e/reporters/progress.go#L70-L71

and the comments for that flag,
https://github.com/kubernetes/kubernetes/blob/2bb77a13b1a709e0208c6ba9ba5323a5e54f79d6/test/e2e/framework/test_context.go#L351


the conclusion from me is that this reporter cannot be disabled manually, it's enabled by default, but it's only matter if the flag `progress-report-url` is set, otherwise, there will be no msg sent out.

aojea replied:
yeah, well, it seems I phrased it wrong , but you answer my question

Thanks

pohly replied:
The reporter used to have the side effect of writing JSON messages to stdout. But I removed that in 1.26.

Now it is as @chendave said, mostly the presence of this reporter can be ignored during normal test runs.

pohly replied:
@chendave: does this get called on all worker processes or only on one?

I have asked a few times whether sonobuoy uses this report when running tests in parallel, but so far we have only been able to determine that sonobuoy supports parallel testing. It remains unclear to me whether that is getting used.

My suspicion is that this report is going to be broken when running tests in parallel because each process is going to report its own, local process to the remote URL, with no way of consolidating the different (and conflicting) reports.

We can't fix this, but perhaps you can check and document this as a known limitation?

chendave replied:
@pohly  I tried to use `tcpdump` to dump the data package, 5 testcases are executed / 3 procs / 3 failed and 2 passed. 

here is the data I captured.

```
{"msg":"","total":5,"completed":0,"skipped":0,"failed":0}
{"msg":"Test Suite starting","total":5,"completed":0,"skipped":0,"failed":0}
{"msg":"FAILED [sig-node] NoExecuteTaintManager Single Pod doesn't evict pod with tolerations from tainted nodes","total":0,"completed":0,"skipped":1332,"failed":1,"failures":["[sig-node] NoExecuteTaintManager Single Pod doesn't evict pod with tolerations from tainted nodes"]}
{"msg":"FAILED [sig-node] NoExecuteTaintManager Single Pod evicts pods from tainted nodes","total":5,"completed":0,"skipped":2524,"failed":1,"failures":["[sig-node] NoExecuteTaintManager Single Pod evicts pods from tainted nodes"]}
{"msg":"PASSED [sig-node] NoExecuteTaintManager Single Pod removing taint cancels eviction [Disruptive] [Conformance]","total":0,"completed":1,"skipped":1181,"failed":0}
{"msg":"Test Suite completed","total":0,"completed":1,"skipped":2592,"failed":0}
{"msg":"PASSED [sig-node] NoExecuteTaintManager Single Pod eventually evict pod with finite tolerations from tainted nodes","total":0,"completed":1,"skipped":1399,"failed":1,"failures":["[sig-node] NoExecuteTaintManager Single Pod doesn't evict pod with tolerations from tainted nodes"]}
{"msg":"Test Suite completed","total":0,"completed":1,"skipped":1399,"failed":1,"failures":["[sig-node] NoExecuteTaintManager Single Pod doesn't evict pod with tolerations from tainted nodes"]}
{"msg":"FAILED [sig-node] NoExecuteTaintManager Single Pod pods evicted from tainted nodes have pod disruption condition","total":5,"completed":0,"skipped":3076,"failed":2,"failures":["[sig-node] NoExecuteTaintManager Single Pod evicts pods from tainted nodes","[sig-node] NoExecuteTaintManager Single Pod pods evicted from tainted nodes have pod disruption condition"]}
{"msg":"Test Suite completed","total":5,"completed":0,"skipped":3076,"failed":2,"failures":["[sig-node] NoExecuteTaintManager Single Pod evicts pods from tainted nodes","[sig-node] NoExecuteTaintManager Single Pod pods evicted from tainted nodes have pod disruption condition"]}
```

the reporter before the suite is started is correct.
```
{"msg":"Test Suite starting","total":5,"completed":0,"skipped":0,"failed":0}
```

but yes, each proc will report on it's own and here are three items reported after the suite is finished, `Test Suite completed`.

chendave replied:
> does this get called on all worker processes or only on one?

for the msg print, there is only one item for `ReportBeforeSuite`.

pohly replied:
As I suspected, the messages from the additional workers don't have the right `total`. The other stats in "Test Suite completed" are also incomplete (`completed` is the number of completed tests in that worker, not in the suite).

I don't think that this is a regression, because I don't see how this could have worked differently before with Ginkgo v1, but I am not sure. Before we proceed it would be good to get confirmation from someone familiar with sonobuoy that this behavior is okay. Even better would be a unit test in Kubernetes itself. This could be similar to how I am testing output handling (see https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/log_test.go), except that the unit test needs to be in a separate directory similar to the [cleanup test](https://github.com/kubernetes/kubernetes/tree/master/test/e2e/framework/internal/unittests/cleanup) because defining the Ginkgo test suite is global.

@raosaurabh: does this fix the issue?

/cc @BobyMCbobs

chendave replied:
> completed is the number of completed tests in that worker, not in the suite

take a look at those msg you will find that if the test passed, the `completed` is set correctly and if it is failed it will be set as *0* instead.

> I don't think that this is a regression

+1

> Even better would be a unit test in Kubernetes itself.

unit test will be great, but I think it's not easy to achieve with an unit test, how do we launch a webservice to response the post request within a UT, and do we really need to test a questionable feature provided by ginkgo.
I will take an investigation with regarding to the unittest but I think it's only nice to have as long as this PR fixed that issue https://github.com/kubernetes/kubernetes/issues/114313

pohly replied:
Let me prepare a unit test with some TODOs - shouldn't take long, I know that code.

chendave replied:
> completed is the number of completed tests in that worker, not in the suite

seems like it is calculated as this from `Test Suite completed`, total is 5.
proc1:  1 completed (passed)
proc2:  1 completed (passed), 1 failed
proc3:   2 failed 

this matches with the total test that are executed. `total` is not set correctly, only proc3 has the value set but it's for the entire suite.

pohly replied:
See https://github.com/chendave/kubernetes/pull/2

pohly replied:
It cannot run Ginkgo tests in parallel, though.

onsi replied:
for what it's worth - the behavior y'all are noticing isn't a regression.  it was always broken when running in parallel.  options to fix would include:
- having the server do the aggregation
- spinning up a server on the first parallel process and having other processes report to it, then it report to the central server
- open an issue on ginkgo and ask me how i'd support this (i'd probably introduce `CurrentSuiteReport()` which would allow you to get a synchronized view of the entire suite at any point from any running node.

chendave replied:
@pohly I think we'd better to create a separate pr for the UT bit, that doesn't seems like related with the issue we would like to fix.

I'm going to reset unless you disagree.

pohly replied:
That's okay.

aojea commented:
/lgtm

based on https://github.com/kubernetes/kubernetes/pull/114417#discussion_r1055235897, but defer to Patrick , he is more familiar with the technical details, he must approve

/assign @pohly

chendave committed:
e2e: comment the known limitation of the `ProgressReporter`

Signed-off-by: Dave Chen <dave.chen@arm.com>

chendave commented:
@pohly  I add another commit to comment the known limitation since the original commit is to bring the total test spec back, and actually, the limitation doesn't impact that fix at all. So I add another commit to the comment `ProgressReporter` instead. Pls let me know if you prefer a squash or not.

pohly commented on test/e2e/reporters/progress.go:
```diff
@@ -33,6 +33,11 @@ import (
+// One known limitation of the ProgressReporter it this reporter will not consolidate the reporter from each sub-process
+// if the tests are run in parallel.
+// As what's observed the reporter sent before test suite is started will be assembled correctly but each process will report
+// on its own after each test or suite are completed.
```
Perhaps include the message from your TCP dump as example here?

chendave replied:
Done

chendave commented:
/test pull-kubernetes-unit

chendave commented on test/e2e/reporters/progress_test.go:
```diff
@@ -0,0 +1,122 @@
+*/
+
+// This test uses etcd that is only fully supported for AMD64 and Linux
+// https://etcd.io/docs/v3.5/op-guide/supported-platform/#support-tiers
```
we need to remove these line.

chendave commented:
We haven't heard from @raosaurabh and @BobyMCbobs for quite a while.

Can we get this fix in or we still need to wait for some feedback? 

@pohly @aojea

pohly commented:
Let's merge it.

/lgtm
/approve

chendave commented:
/test pull-kubernetes-e2e-gce

dims commented:
@pohly @chendave do we want to cherry-pick this to older branches?

pohly commented:
This needs a ginkgo update - backporting will be difficult. I would only try to backport if there's strong evidence that it is needed - so far we don't even have the confirmation that it really fixes the problem.
